eng/scripts/distribute-packages-to-matrix.ps1

```diff
@@ -0,0 +1,233 @@
+        # is a different property. We do this because we want to ensure that the matrix is fully covered, even if
+        # we are generating the matrix sparsely, we still want full coverage of these targetingStrings
+        $targetingStringArray = @($targetingString) * $MatrixMultiplier
+
+        if ($matrixUpdate) {
+            $matrix.matrix.ArtifactPackageNames += $targetingStringArray
+        }
+
+        # if there were any include objects, we need to duplicate them exactly and add the targeting string to each
+        # this means that the number of includes at the end of this operation will be incoming # of includes * the number of batches
+        if ($includeCount -gt 0) {
+            $includeCopy = $OriginalIncludeObject | ConvertTo-Json -Depth 100 | ConvertFrom-Json
+
+            Update-Include -Matrix $matrix -IncludeConfig $includeCopy -TargetingString $targetingString
+        }
+    }
+
+    foreach($batch in $IndirectBatches) {
+        $targetingString = ($batch | Select-Object -ExpandProperty Name) -join ","
+        if ($matrixUpdate) {
+            $matrix.matrix.ArtifactPackageNames += @($targetingString)
+        }
+    }
+
+}
+
+# calculate general targeting information and create our batches prior to updating any matrix
+$packageProperties = Get-ChildItem -Recurse "$PackageInfoFolder" *.json `
+    | % { Get-Content -Path $_.FullName | ConvertFrom-Json }
+
+# I will assign all the direct included packages first. our goal is to get full coverage of the direct included packages
+# then, for the indirect packages, we will add them as sparse ArtifactPackageNames bundles to the matrix
+$batches = Split-ArrayIntoBatches -InputArray $packageProperties -BatchSize $BATCHSIZE
+
+$matrix = Get-Content -Path $PlatformMatrix | ConvertFrom-Json
+
+$matrixLocation = $PlatformMatrix
+
+# handle the case where we discover an import, we need to climb the dependency in that case
+if ($matrix.matrix.PSObject.Properties["`$IMPORT"]) {
+    $originalMatrix = $matrix
+    $originalMatrixLocation = $PlatformMatrix
+
+    # rest of the update will happen to that matrix object
+    $matrixLocation = (Join-Path $RepoRoot $matrix.matrix."`$IMPORT")
+    $matrix = Get-Content -Path $matrixLocation | ConvertFrom-Json
+
+    # we just need to walk the include objects and update the targeting string for the batches we have before
+    # updating in place, then continuing on to update the actual matrix object that we imported
+    # there is a clear and present edge case here with nested import. We will not handle that case
+
+    $originalInclude = $originalMatrix.include
+    $originalMatrix.include = @()
+
+    # update the include objects for the original matrix that was importing
+    Update-Matrix -Matrix $originalMatrix -DirectBatches $directBatches -IndirectBatches @() -MatrixMultiplier 1 -IncludeCount $matrix.include.Count -OriginalIncludeObject $originalInclude
```

Comment:
I don't have a full understanding yet of how this works, but would it be simpler to generate the matrix first via eng/common scripts or the job template so that all includes/imports/etc. are processed as a baseline. Then we only have to process the outputted matrix which doesn't have any of the "special" fields and simplifies the logic here.

I worry with all this post-processing that any changes to the underlying generator could break downstream logic in these repo scripts. It feels a bit too coupled.

Reply:
Similarly, for things above like extract multiplier, are there places we could simplify by moving logic into the core matrix generation scripts? That way we can re-use some of this stuff across languages.

Reply:
> I don't have a full understanding yet of how this works, but would it be simpler to generate the matrix first via eng/common scripts or the job template so that all includes/imports/etc. are processed as a baseline. Then we only have to process the outputted matrix which doesn't have any of the "special" fields and simplifies the logic here.
> 
> I worry with all this pre-processing that any changes to the underlying generator could break downstream logic in these repo scripts. It feels a bit too coupled.

Ok. My thought is we add the concept of `PostGenerationSteps` that can manipulate the matrix further. This transitions the logic to

> either set property on each item in the list of matrix configs, or duplicate if necessary for additional batches